libs/libc/machine/arm64/arch_elf.c

```diff
@@ -0,0 +1,789 @@
+
+  return 0;
+}
+
+static int reloc_data(enum reloc_op_e op, uintptr_t place,
+                      uint64_t val, int len)
+{
+  int64_t sval = do_reloc(op, place, val);
+
+  /* The ELF psABI for AArch64 documents the 16-bit and 32-bit place
+   * relative and absolute relocations as having a range of [-2^15, 2^16)
+   * or [-2^31, 2^32), respectively. However, in order to be able to
+   * detect overflows reliably, we have to choose whether we interpret
+   * such quantities as signed or as unsigned, and stick with it.
+   * The way we organize our address space requires a signed
+   * interpretation of 32-bit relative references, so let's use that
+   * for all R_AARCH64_PRELxx relocations. This means our upper
+   * bound for overflow detection should be Sxx_MAX rather than Uxx_MAX.
+   */
+
+  switch (len)
+    {
+      case 16:
+        {
+          *(int16_t *)place = sval;
+          switch (op)
+            {
+              case RELOC_OP_ABS:
+                {
+                  if (sval < 0 || sval > UINT16_MAX)
+                    {
+                      return -ERANGE;
+                    }
+                }
+                break;
+
+              case RELOC_OP_PREL:
+                {
+                  if (sval < INT16_MIN || sval > INT16_MAX)
+                    {
+                      return -ERANGE;
+                    }
+                }
+                break;
+
+              default:
+                {
+                  berr("Invalid 16-bit data relocation (%d)\n", op);
+                  return -EINVAL;
+                }
+            }
+        }
+        break;
+
+      case 32:
+        {
+          *(int32_t *)place = sval;
+          switch (op)
+            {
+              case RELOC_OP_ABS:
+                {
+                  if (sval < 0 || sval > UINT32_MAX)
+                    {
+                      return -ERANGE;
+                    }
+                }
+                break;
+
+              case RELOC_OP_PREL:
+                {
+                  if (sval < INT32_MIN || sval > INT32_MAX)
+                    {
+                      return -ERANGE;
+                    }
+                }
+                break;
+
+              default:
+                {
+                  berr("Invalid 32-bit data relocation (%d)\n", op);
+                  return -EINVAL;
+                }
+            }
+        }
+        break;
+
+      case 64:
+        {
+          *(int64_t *)place = sval;
+        }
+        break;
+
+      default:
+        {
+          berr("Invalid length (%d) for data relocation\n", len);
+          return -EINVAL;
+        }
+    }
+
+  return 0;
+}
+
+static int reloc_insn_movw(enum reloc_op_e op, uintptr_t place,
+                           uint64_t val, int lsb,
+                           enum insn_movw_imm_type_e imm_type)
+{
+  uint32_t insn = htole32(*(uint32_t *)place);
+  uint64_t imm;
+  int64_t sval;
+
+  sval = do_reloc(op, place, val);
+  imm = sval >> lsb;
```

Comment:
Just to confirm that usage of a signed type (sing bit propagation) is used intentionally here?

Reply:
Yes, follow linux. ditto.